src/passes/ReorderGlobals.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Sorts globals by their static use count. This helps reduce the size of wasm
+// binaries because fewer bytes are needed to encode references to frequently
+// used globals.
+//
+
+#include "memory"
+
+#include "ir/find_all.h"
+#include "pass.h"
+#include "wasm.h"
+
+namespace wasm {
+
+using NameCountMap = std::unordered_map<Name, std::atomic<Index>>;
+
+struct UseCountScanner : public WalkerPass<PostWalker<UseCountScanner>> {
+  bool isFunctionParallel() override { return true; }
+
+  bool modifiesBinaryenIR() override { return false; }
+
+  UseCountScanner(NameCountMap& counts) : counts(counts) {}
+
+  std::unique_ptr<Pass> create() override {
+    return std::make_unique<UseCountScanner>(counts);
+  }
+
+  void visitGlobalGet(GlobalGet* curr) {
+    // We can't add a new element to the map in parallel.
+    assert(counts.count(curr->name) > 0);
+    counts[curr->name]++;
+  }
+  void visitGlobalSet(GlobalSet* curr) {
+    assert(counts.count(curr->name) > 0);
+    counts[curr->name]++;
+  }
+
+private:
+  NameCountMap& counts;
+};
+
+struct ReorderGlobals : public Pass {
+  // Whether to always reorder globals, even if there are very few and the
+  // benefit is minor. That is useful for testing.
+  bool always;
+
+  ReorderGlobals(bool always) : always(always) {}
+
+  void run(Module* module) override {
+    if (module->globals.size() < 128 && !always) {
+      // The module has so few globals that they all fit in a single-byte U32LEB
+      // value, so no reordering we can do can actually decrease code size. Note
+      // that this is the common case with wasm MVP modules where the only
+      // globals are typically the stack pointer and perhaps a handful of others
+      // (however, features like wasm GC there may be a great many globals).
+      return;
+    }
+
+    NameCountMap counts;
+    // Fill in info, as we'll operate on it in parallel.
+    for (auto& global : module->globals) {
+      counts[global->name];
+    }
+
+    // Count uses.
+    UseCountScanner scanner(counts);
+    scanner.run(getPassRunner(), module);
+    scanner.runOnModuleCode(getPassRunner(), module);
+
+    // To sort, we must first find dependencies, since if $b's definition
```

Comment:
I think this could be much simpler if you use the `TopologicalSort` utility. If you sort the initial items from least used to most used, then the topological sort will generally order them with the most used items first. We use this strategy to sort types by decreasing use here: https://github.com/WebAssembly/binaryen/blob/main/src/ir/module-utils.cpp#L297-L300

Reply:
Good idea, done. It's not less code, it turns out, but it is simpler and avoids slowdowns.